File: Java/041.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Problem041 {

    // Sieve for primality testing up to a certain limit
    private static boolean[] isPrimeSieve;
    private static int SIEVE_LIMIT = 7654321; // Max 7-digit pandigital number

    public static void sievePrimes() {
        isPrimeSieve = new boolean[SIEVE_LIMIT + 1];
        Arrays.fill(isPrimeSieve, true);
        isPrimeSieve[0] = false;
        isPrimeSieve[1] = false;
        for (int p = 2; p * p <= SIEVE_LIMIT; p++) {
            if (isPrimeSieve[p]) {
                for (int i = p * p; i <= SIEVE_LIMIT; i += p) {
                    isPrimeSieve[i] = false;
                }
            }
        }
    }

    public static boolean isPrime(int n) {
        if (n < 0 || n > SIEVE_LIMIT) {
            // Fallback for numbers outside sieve range, though not expected for this problem
            if (n <= 1) return false;
            if (n <= 3) return true;
            if (n % 2 == 0 || n % 3 == 0) return false;
            for (int i = 5; i * i <= n; i = i + 6) {
                if (n % i == 0 || n % (i + 2) == 0) {
                    return false;
                }
            }
            return true;
        }
        return isPrimeSieve[n];
    }

    public static boolean isPandigital(int n) {
        String s = String.valueOf(n);
        int length = s.length();
        Set<Character> digits = new HashSet<>();

        for (char c : s.toCharArray()) {
            if (c == '0') return false; // Pandigital numbers don't contain 0
            digits.add(c);
        }

        if (digits.size() != length) return false; // Check for unique digits

        for (int i = 1; i <= length; i++) {
            if (!digits.contains(Character.forDigit(i, 10))) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        sievePrimes();

        // We only need to check 7-digit pandigital numbers.
        // 9-digit pandigital numbers (1-9) sum of digits is 45, divisible by 3.
        // 8-digit pandigital numbers (1-8) sum of digits is 36, divisible by 3.
        // So, no 9-digit or 8-digit pandigital numbers can be prime.
        // The largest possible pandigital prime must be 7-digit or less.
        // The largest 7-digit pandigital number is 7654321.
        for (int i = 7654321; i >= 1; i--) {
            if (isPandigital(i) && isPrime(i)) {
                System.out.println(i);
                break;
            }
        }
    }
}

File: Rust/041.rs
use std::collections::HashSet;

fn sieve_of_eratosthenes(limit: usize) -> HashSet<u32> {
    let mut is_prime = vec![true; limit + 1];
    if limit >= 0 { is_prime[0] = false; }
    if limit >= 1 { is_prime[1] = false; }

    for p in 2..=limit {
        if p * p > limit { // Optimization: only iterate up to sqrt(limit)
            break;
        }
        if is_prime[p] {
            for multiple in (p * p..=limit).step_by(p) {
                is_prime[multiple] = false;
            }
        }
    }

    let mut primes = HashSet::new();
    for (i, &prime) in is_prime.iter().enumerate() {
        if prime {
            primes.insert(i as u32);
        }
    }
    primes
}

fn is_pandigital(n: u32, num_digits: u32) -> bool {
    let s = n.to_string();
    if s.len() != num_digits as usize {
        return false;
    }

    let mut digits: Vec<char> = s.chars().collect();
    digits.sort_unstable();

    let expected_digits: String = (1..=num_digits).map(|d| d.to_string()).collect();
    digits.into_iter().collect::<String>() == expected_digits
}

fn main() {
    // The sum of digits 1-9 is 45 (divisible by 3).
    // The sum of digits 1-8 is 36 (divisible by 3).
    // So, 9-digit and 8-digit pandigital numbers are divisible by 3.
    // Thus, they cannot be prime (unless the number itself is 3, which is not 8 or 9 digits).
    // We only need to check 7-digit pandigital primes.
    let max_7_digit_pandigital = 7_654_321;
    let primes_set = sieve_of_eratosthenes(max_7_digit_pandigital as usize);

    let mut largest_pandigital_prime = 0;

    // Iterate downwards from the largest possible 7-digit number
    for n in (1_000_000..=max_7_digit_pandigital).rev() {
        if primes_set.contains(&n) && is_pandigital(n, 7) {
            largest_pandigital_prime = n;
            break;
        }
    }

    println!("Largest 7-digit pandigital prime: {}", largest_pandigital_prime);
}
